uitools/toolkitcpp/cpp/Esri/ArcGISRuntime/Toolkit/ArcGISAuthenticationController.cpp

```diff
@@ -0,0 +1,263 @@
+
+namespace Esri::ArcGISRuntime::Toolkit {
+
+/*!
+ \class Toolkit::ArcGISAuthenticationController
+ \inmodule EsriArcGISRuntimeToolkit
+ \ingroup ArcGISQtToolkitUiCppControllers
+ \brief In MVC architecture, this is the controller for the corresponding
+        AuthenticationView.
+ */
+
+/*!
+  \brief Constructor
+  \list
+    \li \a parent Owning QObject.
+  \endlist
+ */
+ArcGISAuthenticationController::ArcGISAuthenticationController(QObject* parent) :
+  ArcGISAuthenticationChallengeHandler(parent)
+{
+  if (!canBeUsed_())
+    return;
+
+  // Set the authentication manager to use this controller
+  ArcGISRuntimeEnvironment::instance()->arcGISAuthenticationManager()->setArcGISAuthenticationChallengeHandler(this);
+
+  // listen for OAuth prompts
+  connect(ArcGISRuntimeEnvironment::instance()->arcGISAuthenticationManager(), &ArcGISAuthenticationManager::oAuthUserLoginPromptIssued, this,
+          [this](OAuthUserLoginPrompt* currentOAuthUserLoginPrompt)
+  {
+    m_currentOAuthUserLoginPrompt = currentOAuthUserLoginPrompt;
+    emit authorizeUrlChanged();
+    emit preferPrivateWebBrowserSessionChanged();
+    emit redirectUrlChanged();
+    emit displayOAuthSignInView();
+  });
+}
+
+/*!
+  \brief Destructor.
+ */
+ArcGISAuthenticationController::~ArcGISAuthenticationController() = default;
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::create(QQmlEngine* qmlEngine, QJSEngine* /*jsEngine*/)
+{
+  static QPointer<ArcGISAuthenticationController> instance = new ArcGISAuthenticationController(qmlEngine);
+
+  if (!instance)
+    instance = new ArcGISAuthenticationController(qmlEngine);
+
+  return instance;
+}
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::instance()
+{
+  return create(nullptr, nullptr);
+}
+
+void ArcGISAuthenticationController::handleArcGISAuthenticationChallenge(ArcGISAuthenticationChallenge* challenge)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  m_currentChallenge = challenge;
+  emit currentAuthenticatingHostChanged();
+
+  // first see if we can handle this with OAuth
+  const auto requestUrl = challenge->requestUrl();
+  for (OAuthUserConfiguration* userConfiguration : std::as_const(m_userConfigurations))
+  {
+    if (userConfiguration->canBeUsedForUrl(requestUrl))
+    {
+      OAuthUserCredential::createAsync(userConfiguration, this).then(this, [this](OAuthUserCredential* credential)
+      {
+        if (!m_currentChallenge)
+          return;
+
+        m_currentChallenge->continueWithCredential(credential);
+        m_currentChallenge->deleteLater();
+        m_currentChallenge = nullptr;
+      }).onFailed(this, [this](const ErrorException& e)
+      {
+        if (!m_currentChallenge)
+          return;
+
+        m_currentChallenge->continueAndFailWithError(e.error());
+        m_currentChallenge->deleteLater();
+        m_currentChallenge = nullptr;
+      });
+
+      return;
+    }
+  }
+
+  emit displayUsernamePasswordSignInView();
```

Comment:
What is the role of this signal? It's not emitted if one user configuration is used for request URL. Is it expected?

Reply:
Currently there are two supported workflows:
1. Token (username/password dialog).
2. OAuth.

Users have to opt-in to OAuth as you know, but we prefer OAuth if users have provided any `OAuthUserConfiguration` so that is checked first. If there are no matching `OAuthUserConfiguration`, then we fall through to username/password dialog - and that's what this signal is for.

>It's not emitted if one user configuration is used for request URL. Is it expected?

Yes, it is expected. `ArcGISAuthenticationManager::oAuthUserLoginPromptIssued` will emit from the API, and the toolkit code above will do `emit displayOAuthSignInView();`.